### PR TITLE
Adding time of impact computation using GJK

### PR DESCRIPTION
**Pull request by Rhuagh**
Also adding distance computation for complex shapes

### CODE REVIEW

File: src/algorithm/minkowski/gjk/mod.rs [hunk 1]
```diff
@@ -115,6 +117,115 @@ where
...
+    /// - `right_transform`: model-to-world-transform for the right primitive
+    ///
+    /// ## Returns:
+    ///
+    /// Will optionally return the time of impact. If no collision was detected, None is returned.
+    #[allow(unused_variables)]
+    pub fn intersect_time_of_impact<P, PL, PR, TL, TR>(
+        &self,
+        left: &PL,
+        left_start_transform: &TL,
```
Comment (kvark) [inline src/algorithm/minkowski/gjk/mod.rs]: let's use ranges there, e.g. `left_transforms: Range<&TL>`

File: src/algorithm/minkowski/gjk/mod.rs [hunk 2]
```diff
@@ -115,6 +117,115 @@ where
...
+        right_end_transform: &TR,
+    ) -> Option<P::Scalar>
+    where
+        P: EuclideanSpace,
+        P::Scalar: BaseFloat,
+        PL: SupportFunction<Point = P>,
+        PR: SupportFunction<Point = P>,
+        SP: SimplexProcessor<Point = P>,
+        P::Diff: Neg<Output = P::Diff> + InnerSpace,
+        TL: Transform<P> + TranslationInterpolate<P::Scalar>,
```
Comment (kvark) [inline src/algorithm/minkowski/gjk/mod.rs]: I'm curious to see the compile times on that :)

File: src/algorithm/minkowski/gjk/mod.rs [hunk 3]
```diff
@@ -115,6 +117,115 @@ where
...
+        let mut lambda = P::Scalar::zero();
+
+        // build the start transforms
+        let mut left_curr_transform =
+            left_start_transform.translation_interpolate(left_end_transform, lambda);
+        let mut right_curr_transform =
+            right_start_transform.translation_interpolate(right_end_transform, lambda);
+
+        // build simplex and get the first support point
+        let mut simplex: Vec<SupportPoint<P>> = Vec::with_capacity(5);
```
Comment (kvark) [inline src/algorithm/minkowski/gjk/mod.rs]: explicit type is not needed here

File: src/algorithm/minkowski/gjk/mod.rs [hunk 4]
```diff
@@ -292,35 +398,168 @@ where
...
+                }
+            }
+        }
+
+        // CollisionOnly handling will have returned already if there was a contact, so this
+        // scenario will only happen when we have a contact in FullResolution mode
+        if contacts.len() > 0 {
+            // penetration depth defaults to 0., and can't be nan from EPA,
+            // so unwrapping is safe
+            contacts.into_iter().max_by(|l, r| {
```
Comment (kvark) [inline src/algorithm/minkowski/gjk/mod.rs]: nit: `max_by` returns an `Option`, so no need to check `contacts.len() > 0` above

File: src/algorithm/minkowski/gjk/mod.rs [hunk 5]
```diff
@@ -292,35 +398,168 @@ where
...
+                    &left_transform,
+                    right_primitive,
+                    &right_transform,
+                ) {
+                    None => return None, // colliding,
+                    Some(distance) => {
+                        min_distance = match min_distance {
+                            None => Some(distance),
+                            Some(min_distance) => {
+                                if distance < min_distance {
```
Comment (kvark) [inline src/algorithm/minkowski/gjk/mod.rs]: `Some(cmp::min(...))` ?

File: src/algorithm/minkowski/gjk/mod.rs [hunk 6]
```diff
@@ -292,35 +398,168 @@ where
...
+    ///
+    /// Will optionally return time of impact if a collision was detected.
+    /// In `CollisionOnly` mode, this contact will only be a time of impact. For `FullResolution`
+    /// mode, the time of impact will be the earliest found among all shape primitives.
+    /// Will return None if no collision was found.
+    pub fn intersect_complex_time_of_impact<P, PL, PR, TL, TR>(
+        &self,
+        strategy: &CollisionStrategy,
+        left: &[(PL, TL)],
+        left_start_transform: &TL,
```
Comment (kvark) [inline src/algorithm/minkowski/gjk/mod.rs]: similarly, `Range` would look great here

File: src/traits.rs [hunk 7]
```diff
@@ -128,3 +128,110 @@ where
...
+        }
+    }
+
+    impl<S> Interpolate<S> for Basis3<S>
+    where
+        S: BaseFloat,
+    {
+        fn interpolate(&self, other: &Self, amount: S) -> Self {
+            Basis3::from(
+                Quaternion::from(*self.as_ref()).lerp(Quaternion::from(*other.as_ref()), amount),
```
Comment (kvark) [inline src/traits.rs]: this doesn't take into account translation/scale?

File: src/traits.rs [hunk 8]
```diff
@@ -128,3 +128,110 @@ where
...
+
+    impl<S> Interpolate<S> for Basis2<S>
+    where
+        S: BaseFloat,
+    {
+        fn interpolate(&self, other: &Self, amount: S) -> Self {
+            // to complex numbers
+            let self_mat = self.as_ref();
+            let other_mat = other.as_ref();
+            let self_c = Vector2::new(self_mat.x.x, self_mat.x.y);
```
Comment (kvark) [inline src/traits.rs]: is it different from `self_mat.x`?

File: src/traits.rs [hunk 9]
```diff
@@ -128,3 +128,110 @@ where
...
+            let other_c = Vector2::new(other_mat.x.x, other_mat.x.y);
+            // do interpolation
+            let c = self_c.lerp(other_c, amount);
+            // to basis
+            Rotation2::from_angle(Rad(c.x.acos()))
```
Comment (kvark) [inline src/traits.rs]: somewhat unfortunate to call `acos` here just to later see calls to `cos/sin`

File: src/traits.rs [hunk 10]
```diff
@@ -128,3 +128,110 @@ where
...
+        }
+    }
+
+    impl<V, R> Interpolate<V::Scalar> for Decomposed<V, R>
```
Comment (kvark) [inline src/traits.rs]: we need a dual quaternion transform representation for fancier interpolation ;)

File: src/traits.rs (hunk 7 above)
Comment (Rhuagh) [reply src/traits.rs]: Basis3 only express a rotation from what I've seen?

File: src/traits.rs (hunk 9 above)
Comment (Rhuagh) [reply src/traits.rs]: Yeah, this is a limitation in cgmath, the only way to create a Basis2 is from an angle :/

File: src/traits.rs (hunk 8 above)
Comment (Rhuagh) [reply src/traits.rs]: Hmm, no, you're right.

File: src/traits.rs (hunk 9 above)
Comment (Rhuagh) [reply src/traits.rs]: If I could just create the Basis2 directly from values I would be home free, but I can't. C'est la vie.

File: src/traits.rs (hunk 7 above)
Comment (Rhuagh) [reply src/traits.rs]: According to the cgmath docs: `A three-dimensional rotation matrix.`

File: src/traits.rs (hunk 8 above)
Comment (Rhuagh) [reply src/traits.rs]: I got confused by the fact I'm storing an imaginary number as a Vector2 :P

File: src/algorithm/minkowski/gjk/mod.rs (hunk 2 above)
Comment (Rhuagh) [reply src/algorithm/minkowski/gjk/mod.rs]: They're not that bad tbh :)

File: src/algorithm/minkowski/gjk/mod.rs (hunk 5 above)
Comment (Rhuagh) [reply src/algorithm/minkowski/gjk/mod.rs]: float does not form a complete ordering, and therefore have no Ord implementation, so can't use std::cmp unfortunately.

File: src/traits.rs (hunk 7 above)
Comment (Rhuagh) [reply src/traits.rs]: I've gone through cgmath code now, and there is no way to make Basis3 any other than a rotation matrix, so there is only rotation in there to interpolate.

**Comment by Rhuagh** [comment]
Review comments addressed or commented on.

**Comment by kvark** [comment]
Sweet, thank you!
bors r+

**Pull request merged**